### PR TITLE
fix instantiation of key types and key_type_from_name

### PR DESCRIPTION
**Pull request by enkore**
- Fixed key_type_from_name
- The KeyType classes were built in a somewhat weird way (though I understand this was probably due to them being cdef and the resulting restrictions, i.e. `_type = ...` simply isn't allowed on a cdef field, so things got awkward). Due to this, just doing "keytypes.RSAKey()" and using that object as a key type would always fail. This is now fixed. E.g. `key.generate(keytypes.RSAKey(), 4096)` works now.
- Added Ed25519 key type.
- Added examples/keygen.py. The resulting keys are directly interoperable with ssh(1) from OpenSSH.

(eref 42d478fb6976)

### CODE REVIEW

**Review by pkittenis — CHANGES_REQUESTED** [review]
Thanks for the interest.

Looks good other than a short test.

**Comment by enkore** [comment]
Good points, I redid the change using cdefs and added a test for this.

File: ssh/keytypes.pyx [hunk 1]
```diff
@@ -31,115 +30,131 @@ cdef class KeyType:
...
 
     def __str__(self):
-        cdef bytes _type
         cdef const_char *c_type
-        with nogil:
-            c_type = ssh_key_type_to_char(self._type)
-        _type = c_type
-        return to_str(_type)
+        c_type = ssh_key_type_to_char(self._type)
+        if c_type:
```
Comment (pkittenis) [inline ssh/keytypes.pyx:35]: Safer as `if c_type is not NULL`

File: ssh/keytypes.pyx [hunk 2]
```diff
@@ -31,115 +30,131 @@ cdef class KeyType:
...
+            return to_str(c_type)
+        else:
```
Comment (pkittenis) [inline ssh/keytypes.pyx:37]: `else` not needed here.

**Review by pkittenis — CHANGES_REQUESTED** [review]

**Comment by pkittenis** [comment]
Looks great, thanks for making these changes.

Couple more minor points in review.

File: ssh/keytypes.pyx (hunk 1 above)
Comment (enkore) [reply ssh/keytypes.pyx]: Certainly a little clearer considering Cython's mixed type system; done

**Comment by pkittenis** [comment]
Thanks for the contribution.

Added another test for keygen which we did not have.

**Pull request merged**